### PR TITLE
Add support for the USB-SD-Mux FAST

### PR DESCRIPTION
**Pull request by SmithChart**
The USB-SD-Mux FAST is a next generation prototype of the USB-SD-Mux.
This new generation contains  a new *I2C GPIO expander* and also provides two external open-drain outputs.

This PR adds support for this new hardware to this tooling.

For co-existence of old and new USB-SD-Muxes this PR also adds support to automatically detect the version of the USB-SD-Mux. This feature will also help to catch cases where the tooling is accidentally used on a another SCSI device (e.g. a hard drive or generic card reader).

### CODE REVIEW

File: usbsdmux/usbsdmux.py [hunk 1]
```diff
@@ -18,12 +18,56 @@
...
 
+import os
 import time
 
-from .pca9536 import Pca9536
+from .i2c_gpio import Pca9536, Tca6408
 
 
-class UsbSdMux(object):
+class UsbSdMuxFactory:
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:27]: Use a normal function instead of a class which is never instantiated and only has one static method.

File: usbsdmux/usbsdmux.py [hunk 2]
```diff
@@ -18,12 +18,56 @@
...
+    SD_MUX_TYPE_LEGACY = "legacy"
+    SD_MUX_TYPE_FAST = "fast"
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:29]: Just use strings or a module level `Enum`.

File: usbsdmux/usbsdmux.py [hunk 3]
```diff
@@ -100,13 +144,104 @@ def mode_host(self, wait=True):
...
+        if val & self._PWR_disable:
+            return "off"
+
+        if val & self._select_DUT:
+            return "dut"
+        return "host"
+
+    def mode_disconnect(self, wait=True):
+        """
+        Will disconnect the Micro-SD Card from both host and DUT.
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:226]: Instead of duplicating the docstrings, put them on the base class.

File: usbsdmux/i2c_gpio.py [hunk 4]
```diff
@@ -97,6 +97,20 @@ def output_values(self, values: int, bitmask: int = 0xFF):
...
+        val = (val & ~bitmask) & 0xFF
+        self._write_register(addr, val)
+
+    def register_set_bit(self, addr: int, bitmask: int):
+        """Sets the bits with '1' in <bitmask> in the register <addr>"""
+
+        val = self.read_register(addr)[0]
+        val = (val | bitmask) & 0xFF
+        self._write_register(addr, val)
+
```
Comment (jluebbe) [inline usbsdmux/i2c_gpio.py:110]: The usual primitive for this is a _update_mask function, which can perform either function.

**Comment by KarlK90** [comment]
@SmithChart I've tested the PR with the new USD-SD-Mux and it works just fine.

File: usbsdmux/i2c_gpio.py (hunk 4 above)
Comment (SmithChart) [reply usbsdmux/i2c_gpio.py:110]: Not sure what you mean by this. Internet search does not report any useful results. Do you have a pointer?

Comment (jluebbe) [reply usbsdmux/i2c_gpio.py:110]: Something like https://elixir.bootlin.com/linux/latest/source/drivers/base/regmap/regmap.c#L3169 and https://elixir.bootlin.com/linux/latest/source/drivers/base/regmap/regmap.c#L3151 ...

So you have `mask` and `val` as input, read the register, reset the mask bits (`tmp = reg & ~mask`), set the bits  in `val` allowed by the mask (`tmp |= val & mask`) and write back to the register. This single function allows updating any bits in a register to new values without touching the others. It can replace your `_reset_bit` and `_set_bit` functions, while also allowing to replace specific bits with a new value consisting of ones and zeros.

File: usbsdmux/usbsdmux.py [hunk 5]
```diff
@@ -18,12 +18,56 @@
...
+            sg_name = os.path.basename(base_sg)
+            try:
+                with open(f"/sys/class/scsi_generic/{sg_name}/device/model", "r") as fh:
+                    model = fh.read().strip()
+                if model == "sdmux HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_LEGACY
+                elif model == "sdFST HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_FAST
+                else:
+                    raise Exception("Could not determine type of USB-SD-Mux")
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:57]: Include the unknown string here.

File: usbsdmux/usbsdmux.py [hunk 6]
```diff
@@ -18,12 +18,56 @@
...
+            # Get the name of the sg-device from the path
+            base_sg = os.path.realpath(sg)
+            sg_name = os.path.basename(base_sg)
+            try:
+                with open(f"/sys/class/scsi_generic/{sg_name}/device/model", "r") as fh:
+                    model = fh.read().strip()
+                if model == "sdmux HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_LEGACY
+                elif model == "sdFST HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_FAST
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:55]: Instead of indirecting via sd_mux_type, you could just return the matched class here. That would also avoid the need to pass a sd_mux_type` to override. You'd only call this function (perhaps called `autoselect`) if you don't know the type already.

That would simplify the control flow a lot.

File: usbsdmux/__main__.py [hunk 7]
```diff
@@ -36,10 +36,19 @@ def main():
...
+        choices=["get", "dut", "client", "host", "off", "gpio0", "gpio1"],
         type=str.lower,
     )
+    parser.add_argument(
+        "gpio_action",
+        help="Action to perform on the GPIOs",
+        nargs="?",
+        default="",
+        choices=["", "low", "0", "high", "1", "get"],
+    )
```
Comment (jluebbe) [inline usbsdmux/__main__.py:51]: Shouldn't this be in a [subparser](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers)? Otherwise, the gpio_action argument would always be accepted, even for the existing "commands".

File: usbsdmux/__main__.py [hunk 8]
```diff
@@ -68,6 +77,19 @@ def main():
...
+        # Preflight checks for GPIO mode
+        if not ctl.type == UsbSdMuxFactory.SD_MUX_TYPE_FAST:
+            print("This USB-SD-Mux does not support GPIOs.", file=sys.stderr)
+            exit(1)
+        if not gpio_action:
+            gpio_action = "get"
+        if mode == "gpio0":
+            gpio = UsbSdMuxFast.gpio0
+        elif mode == "gpio1":
+            gpio = UsbSdMuxFast.gpio1
```
Comment (jluebbe) [inline usbsdmux/__main__.py:92]: `main()` shouldn't need to know about the details of the type. Just try to call methods defined on the base class and handle `NotImplemented` exceptions.

File: usbsdmux/__main__.py [hunk 9]
```diff
@@ -68,6 +77,19 @@ def main():
         print(f"Does {args.sg} really point to an USB-SD-Mux?")
         sys.exit(1)
     mode = args.mode.lower()
+    gpio_action = args.gpio_action.lower()
```
Comment (jluebbe) [inline usbsdmux/__main__.py:80]: Why not extract the number here?

File: usbsdmux/i2c_gpio.py [hunk 10]
```diff
@@ -25,10 +24,10 @@
...
 class I2cGpio(ABC):
     # Registers inside the supported GPIO expanders
-    _register_inputPort = 0x00
-    _register_outputPort = 0x01
-    _register_polarity = 0x02
-    _register_configuration = 0x03
+    register_inputPort = 0x00
+    register_outputPort = 0x01
+    register_polarity = 0x02
+    register_configuration = 0x03
```
Comment (jluebbe) [inline usbsdmux/i2c_gpio.py:30]: This seems to be an unrelated change?

File: usbsdmux/usbsdmux.py [hunk 11]
```diff
@@ -92,6 +92,7 @@ def __init__(self, sg):
         sg -- /dev/sg* to use
         """
         self._pca = Pca9536(sg)
+        self.type = UsbSdMuxFactory.SD_MUX_TYPE_LEGACY
```
Comment (jluebbe) [inline usbsdmux/usbsdmux.py:95]: The class is already a unique object, so there's no need to for this indirection.

**Review by jluebbe — COMMENTED** [review]
In general, why does `read_register` return a list if all callers use `[0]` on the result?

File: usbsdmux/usbsdmux.py (hunk 6 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py:55]: Sorry. I don't understand what you mean here. Is there something lost in the malformed highlight?

**Comment by SmithChart** [comment]
Rebased onto master

File: usbsdmux/usbsdmux.py (hunk 1 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py]: Good point. Changed

File: usbsdmux/usbsdmux.py (hunk 3 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py]: Added a proper base class and moved all comments there.

File: usbsdmux/i2c_gpio.py (hunk 4 above)
Comment (SmithChart) [reply usbsdmux/i2c_gpio.py]: Ah, you mean like the method I had already implemented here: https://github.com/linux-automation/usbsdmux/pull/52/files#diff-11952356576c020ad5be17c7705787e2c06813dd213effeca7e1557ecef29001R89
Removed both methods. Used the existing one. Thx!

**Comment by SmithChart** [comment]
> In general, why does `read_register` return a list if all callers use `[0]` on the result?

Good point. This was introduced in https://github.com/linux-automation/usbsdmux/commit/c36df6096d367cbf81f38940691ef5df869d35d5 . Re-thinking this I have decided to do two things:

* Get rid of all access to device registers outside the drivers. No external user should need to know about the register layout.
* Removed the `len` parameter and just return one register.

File: usbsdmux/usbsdmux.py (hunk 5 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py]: Good point. Also did this for the other exception directly next to it.

File: usbsdmux/usbsdmux.py (hunk 6 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py]: I got rid of the type override and now simply return the correct instance.

File: usbsdmux/i2c_gpio.py (hunk 10 above)
Comment (SmithChart) [reply usbsdmux/i2c_gpio.py]: In the commit it was on: yes.
In general: Access to the registers was used to access some features of the underlying device.

Not I've added some proper abstraction. Register definitions are now private again.

**Comment by SmithChart** [comment]
@jluebbe Addressed your feedback. Please give it another look!

File: usbsdmux/__main__.py [hunk 12]
```diff
@@ -22,24 +22,26 @@
...
+
+    subparsers.add_parser("get", help="Read the current state of the USB-SD-Mux")
+    subparsers.add_parser("dut", help="Switch to the DUT")
+    subparsers.add_parser("client", help="Switch to the DUT")
+    subparsers.add_parser("host", help="Switch to the host")
+    subparsers.add_parser("off", help="Disconnect from host and DUT")
+
+    parser_gpio = subparsers.add_parser("gpio", help="Manipulate a GPIO")
+    parser_gpio.add_argument("gpio", help="The GPIO to change", choices=[0, 1], type=int)
+    parser_gpio.add_argument("action", help="What to do with the GPIO", choices=["low", "0", "high", "1", "get"])
```
Comment (hnez) [inline usbsdmux/__main__.py:44]: Our _GPIOs_ can not be used as inputs, but I think the name is okay nevertheless.
I've been thinking about alternatives but did not come up with a good one yet:

- `output` - too generic
- `gpo` - not as descriptive as gpio, which is a well known term

File: usbsdmux/i2c_gpio.py [hunk 13]
```diff
@@ -88,15 +70,72 @@ def set_pin_to_input(self, pins):
...
+
+
+class Tca6408(I2cGpio):
+    """
+    Interface to control a TCA6408 that is connected to the auxiliary-I2C of a
+    Microchip USB2642.
+    """
+
+    # The TCA6408 I2C slave Address in 7-Bit Format
+    _I2cAddr = 0b0100_000
```
Comment (hnez) [inline usbsdmux/i2c_gpio.py:132]: We are getting into nit-picking territory here, but could you make both addresses `0b…` or `0x…`?
Also having the `_` separator 3 bits in makes sense in i²c 7bit-land but I'd prefer it at a nibble-boundary.

File: usbsdmux/usbsdmux.py [hunk 14]
```diff
@@ -18,16 +19,120 @@
...
+    base_sg = os.path.realpath(sg)
+    sg_name = os.path.basename(base_sg)
+    model_filename = f"/sys/class/scsi_generic/{sg_name}/device/model"
+    try:
+        with open(model_filename, "r") as fh:
+            model = fh.read().strip()
+        if model == "sdmux HS-SD/MMC":
+            return UsbSdMuxLegacy(sg)
+        elif model == "sdFST HS-SD/MMC":
+            return UsbSdMuxFast(sg)
```
Comment (hnez) [inline usbsdmux/usbsdmux.py:49]: `Legacy` has such a bad connotation. The current-gen USB-SD-Muxes are still fine, no need to get rid of them because they are `Legacy`.
And I am also not a huge fan of the new `FAST` name, but I think it is what it is now.

You can mark this as resolved. I just need this comment to link to it whenever I complain about the name :D.

File: usbsdmux/usbsdmux.py [hunk 15]
```diff
@@ -97,16 +180,131 @@ def mode_DUT(self, wait=True):
...
         self._pca.output_values(self._DAT_enable | self._PWR_enable | self._select_HOST | self._card_inserted)
+
+    def gpio_get(self, gpio):
+        raise NotImplementedError()
+
+    def gpio_set_high(self, gpio):
+        raise NotImplementedError()
+
+    def gpio_set_low(self, gpio):
+        raise NotImplementedError()
```
Comment (hnez) [inline usbsdmux/usbsdmux.py:198]: Maybe have the `UsbSdMux` base class raise NotImplementedError for all of its methods and just not implement them in this derived class?

File: usbsdmux/usbsdmux.py [hunk 16]
```diff
@@ -97,16 +180,131 @@ def mode_DUT(self, wait=True):
...
+
+        # If the SD-Card is disabled, we do not need to check for the selected mode.
+        # PWR_disable and DAT_disable are always switched at the same time.
+        # Let's assume it is sufficient to check one of both.
+        if val & self._PWR_disable:
+            return "off"
+
+        if val & self._select_DUT:
+            return "dut"
+        return "host"
```
Comment (hnez) [inline usbsdmux/usbsdmux.py:247]:
```suggestion
            return "dut"

        return "host"
```

**Review by hnez — COMMENTED** [review]
Nothing deal-breaking, mainly nit-picking and opinions.

File: usbsdmux/usbsdmux.py (hunk 15 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py:198]: With the `` @abc.abstractmethod`` decorator on the base class you are not able to create an object from the derived class when not all of these methods have been overridden.
It's not quite the same, but still close :-)
Feel free to re-open.

File: usbsdmux/usbsdmux.py (hunk 14 above)
Comment (jluebbe) [reply usbsdmux/usbsdmux.py:49]:
```suggestion
        if model == "sdmux HS-SD/MMC":
            return UsbSdMuxClassic(sg)
        elif model == "sdFST HS-SD/MMC":
            return UsbSdMuxFast(sg)
```

File: usbsdmux/__main__.py (hunk 12 above)
Comment (SmithChart) [reply usbsdmux/__main__.py:44]: What about describing it like this:
```
    parser_gpio = subparsers.add_parser("gpio", help="Manipulate a GPIO (open drain output only)")
```

This way this would be reflected in the ``--help``:
```
./venv/bin/usbsdmux --help
usage: usbsdmux [-h] SG {get,dut,client,host,off,gpio} ...

positional arguments:
  SG                    /dev/sg* to use
  {get,dut,client,host,off,gpio}
                        Supply one of the following commands to interact with the device
    get                 Read the current state of the USB-SD-Mux
    dut                 Switch to the DUT
    client              Switch to the DUT
    host                Switch to the host
    off                 Disconnect from host and DUT
    gpio                Manipulate a GPIO (open drain output only)

optional arguments:
  -h, --help            show this help message and exit
```

Comment (hnez) [reply usbsdmux/__main__.py:44]: Sounds good. That should avoid confusion for users that find the sub-command and its `get` option by exploring the tool rather than via the manual.

File: usbsdmux/usbsdmux.py (hunk 15 above)
Comment (jluebbe) [reply usbsdmux/usbsdmux.py:198]: There is some benefit in having it explicitly as not implemented in the derived class.

Comment (SmithChart) [reply usbsdmux/usbsdmux.py:199]: @jluebbe Than I am not sure how you want to implement this. What `abc` decoration would you suggest?

Comment (jluebbe) [reply usbsdmux/usbsdmux.py:199]: I meant that we can leave it as is.

**Review by jluebbe — APPROVED** [review]

File: usbsdmux/usbsdmux.py (hunk 14 above)
Comment (SmithChart) [reply usbsdmux/usbsdmux.py]: Renamed `UsbSdMuxLegacy` to `UsbSdMuxClassic`.

**Review by hnez — APPROVED** [review]
